From 6eca69b243202ba5d6b427ff9a45f9b45b652576 Mon Sep 17 00:00:00 2001
From: Yann Gautier <yann.gautier@foss.st.com>
Date: Thu, 5 Jan 2023 18:01:51 +0100
Subject: [PATCH 241/841] feat(st): add an API to retrieve STM32 image header

Add an helper function stm32_get_header_address() to be used
in get_rotpk_hash().
For the moment it just returns STM32MP_HEADER_BASE.
This macro is then defined for STM32MP13 and STM32MP15.
The macro is not directly used in get_rotpk_hash(), as the header
address will change depending on the use case.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git e7a735dd0cdaa66a67ad925b08ac71102fc4645e feat(st): add an API to retrieve STM32 image header]
Signed-off-by: Yann Gautier <yann.gautier@foss.st.com>
Change-Id: Ib78a8cfac9b7514229d654e4d71651d7c33ac5dc
---
 plat/st/common/include/stm32mp_common.h | 2 ++
 plat/st/common/stm32mp_trusted_boot.c   | 5 +----
 plat/st/stm32mp1/stm32mp1_def.h         | 4 ++++
 plat/st/stm32mp1/stm32mp1_private.c     | 5 +++++
 4 files changed, 12 insertions(+), 4 deletions(-)

diff --git a/plat/st/common/include/stm32mp_common.h b/plat/st/common/include/stm32mp_common.h
index d470f1075..bd6c433df 100644
--- a/plat/st/common/include/stm32mp_common.h
+++ b/plat/st/common/include/stm32mp_common.h
@@ -86,6 +86,8 @@ static inline void stm32mp_setup_early_console(void)
 }
 #endif
 
+uintptr_t stm32_get_header_address(void);
+
 bool stm32mp_skip_boot_device_after_standby(void);
 bool stm32mp_is_wakeup_from_standby(void);
 
diff --git a/plat/st/common/stm32mp_trusted_boot.c b/plat/st/common/stm32mp_trusted_boot.c
index 6d89290ef..94716d8a7 100644
--- a/plat/st/common/stm32mp_trusted_boot.c
+++ b/plat/st/common/stm32mp_trusted_boot.c
@@ -22,8 +22,6 @@
 #include <boot_api.h>
 #include <platform_def.h>
 
-#define HEADER_AND_EXT_TOTAL_SIZE 512
-
 static uint8_t der_sha256_header[] = {0x30, 0x31, 0x30, 0x0d, 0x06, 0x09, 0x60,
 	0x86, 0x48, 0x01, 0x65, 0x03, 0x04, 0x02, 0x01, 0x05, 0x00, 0x04, 0x20};
 static uint8_t root_pk_hash[HASH_DER_LEN];
@@ -97,8 +95,7 @@ static int get_rotpk_hash(void *cookie, uint8_t *hash, size_t len)
 	uint32_t pk_idx = 0U;
 	uint8_t calc_hash[BOOT_API_SHA256_DIGEST_SIZE_IN_BYTES];
 	uint8_t otp_hash[BOOT_API_SHA256_DIGEST_SIZE_IN_BYTES];
-	boot_api_image_header_t *hdr = (boot_api_image_header_t *)(SRAM3_BASE + SRAM3_SIZE -
-								   HEADER_AND_EXT_TOTAL_SIZE);
+	boot_api_image_header_t *hdr = (boot_api_image_header_t *)stm32_get_header_address();
 	boot_extension_header_t *ext_header = (boot_extension_header_t *)hdr->ext_header;
 	boot_ext_header_params_authentication_t *param;
 
diff --git a/plat/st/stm32mp1/stm32mp1_def.h b/plat/st/stm32mp1/stm32mp1_def.h
index b09688a57..c80d456a7 100644
--- a/plat/st/stm32mp1/stm32mp1_def.h
+++ b/plat/st/stm32mp1/stm32mp1_def.h
@@ -149,6 +149,8 @@ enum ddr_type {
 #if STM32MP13
 /* 512 Octets reserved for header */
 #define STM32MP_HEADER_RESERVED_SIZE	U(0x200)
+#define STM32MP_HEADER_BASE		(SRAM3_BASE + SRAM3_SIZE -	\
+					 STM32MP_HEADER_RESERVED_SIZE)
 
 #define STM32MP_BINARY_BASE		STM32MP_SEC_SYSRAM_BASE
 
@@ -158,6 +160,8 @@ enum ddr_type {
 #define STM32MP_PARAM_LOAD_SIZE		U(0x00002400)	/* 9 KB for param */
 /* 256 Octets reserved for header */
 #define STM32MP_HEADER_SIZE		U(0x00000100)
+#define STM32MP_HEADER_BASE		(STM32MP_SEC_SYSRAM_BASE +	\
+					 STM32MP_PARAM_LOAD_SIZE)
 /* round_up(STM32MP_PARAM_LOAD_SIZE + STM32MP_HEADER_SIZE, PAGE_SIZE) */
 #define STM32MP_HEADER_RESERVED_SIZE	U(0x3000)
 
diff --git a/plat/st/stm32mp1/stm32mp1_private.c b/plat/st/stm32mp1/stm32mp1_private.c
index db40a156b..74b09d460 100644
--- a/plat/st/stm32mp1/stm32mp1_private.c
+++ b/plat/st/stm32mp1/stm32mp1_private.c
@@ -281,6 +281,11 @@ void stm32mp1_deconfigure_uart_pins(void)
 }
 #endif
 
+uintptr_t stm32_get_header_address(void)
+{
+	return STM32MP_HEADER_BASE;
+}
+
 uint32_t stm32mp_get_chip_version(void)
 {
 #if STM32MP13
-- 
2.47.2

